### PR TITLE
Refactor `gitBinaryMaskParam` function

### LINKED ISSUE #2465: Refactor how we process handlers params
### What should be done?

The `internal/handlers/common` package contains code responsible for processing and validating various handlers' parameters. This code includes both generic functions such as `GetRequiredParam` and `GetOptionalParam`, as well as parameter-specific functions such as `GetSkipParam` and `GetLimitParam`. However, there are a few issues with this code.

Firstly, the **errors** returned by these functions are not adequately documented. It is unclear what these errors represent and when they should be returned. Additionally, the functions themselves lack documentation about the potential errors they may return.

Secondly, the **functions** are spread out across the entire package, making it challenging to locate and understand them.

To address these issues, significant refactoring is required. In this issue, let's create a separate package `commonparams` dedicated solely to parameter processing. This would consolidate all related code in one place and enable better documentation of both the functions and the errors they may return.

**For the scope of this issue**, the goal is to learn about parameter processing issues and send a pull request with better structuring, documentation, and possibly other improvements. 

### Where?

https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/common/params.go
https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/common/limit.go
https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/common/scale.go
https://github.com/FerretDB/FerretDB/blob/main/internal/handlers/common/skip.go

Check other `Get*Param` functions too.

### Definition of Done

- all handlers updated;
- unit tests added/updated if needed;
- integration/compatibility tests added/updated if needed;
- spot refactorings done;
- significant refactoring are not done.

### PR DESCRIPTION
**Pull request by w84thesun**
# Description

Closes #2465.

## Readiness checklist

* [ ] I added/updated unit tests.
* [x] I added/updated integration/compatibility tests.
* [x] I added/updated comments and checked rendering.
* [x] I made spot refactorings.
* [ ] I updated user documentation.
* [ ] I ran `task all`, and it passed.
* [x] I ensured that PR title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

File: internal/handlers/common/params.go [hunk 1]
```diff
@@ -186,7 +204,14 @@ func GetSkipStageParam(value any) (int64, error) {
 
 // getBinaryMaskParam matches value type, returning bit mask and error if match failed.
 // Possible values are: position array ([1,3,5] == 010101), whole number value and types.Binary value.
-func getBinaryMaskParam(mask any) (uint64, error) {
+//
+// Returned errors:
+//   - ErrBadValue if positional argument is not a number;
```
Comment (chilagrow) [inline internal/handlers/common/params.go:209]:
```suggestion
//   - ErrBadValue if positional argument is not int32;
```

File: internal/handlers/common/params.go [hunk 2]
```diff
@@ -186,7 +204,14 @@ func GetSkipStageParam(value any) (int64, error) {
...
+//   - ErrBadValue if positional argument is negative;
+//   - ErrBadValue if bitmask argument is not a whole number;
```
Comment (chilagrow) [inline internal/handlers/common/params.go:211]:
```suggestion
```
I think we are not returning `ErrBadValue` for non-whole number. We return `ErrFailedToParse`

**Review by chilagrow — COMMENTED** [review]
Great doc :muscle:

Just a couple of minor comments :hugs:

File: internal/handlers/common/params.go [hunk 3]
```diff
@@ -26,8 +26,11 @@ import (
...
 
-// GetRequiredParam returns doc's value for key
-// or protocol error for missing key or invalid type.
+// GetRequiredParam returns doc's value for key.
+//
+// Returned errors:
+//   - ErrBadValue if the document value is missing;
+//   - ErrBadValue if the document value is not of the expected type.
 func GetRequiredParam[T types.Type](doc *types.Document, key string) (T, error) {
 	var zero T
```
Comment (AlekSi) [inline internal/handlers/common/params.go:35]: `defer contract.EnsureError(err, ErrBadValue)`

File: internal/handlers/common/params.go (hunk 1 above)
Comment (w84thesun) [reply internal/handlers/common/params.go]: Fixed. It could be `[int, long, double]`.

File: internal/handlers/common/params.go (hunk 2 above)
Comment (w84thesun) [reply internal/handlers/common/params.go]: Removed that line.

File: internal/handlers/commonparams/params.go [hunk 4]
```diff
@@ -76,8 +72,11 @@ func GetWholeNumberParam(value any) (int64, error) {
 
 // getWholeParamStrict validates the given value for find and count commands.
 //
-// If the value is valid, it returns its int64 representation,
-// otherwise it returns a command error with the given command being mentioned.
+// # If the value is valid, it returns its int64 representation
+//
+// Command errors:
+//   - ErrTypeMismatch if the value is not of types [int, long, double, null];
```
[the remark this answers is not in the document]
Comment (w84thesun) [reply internal/handlers/commonparams/params.go:78]: It return ErrTypeMismatch if the value is `[int, long, double, null]`, so it seems to be correct statement 😄

[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/commonparams/params.go:78]: Just want to double check that I'm not misleading this :eyes:  
https://github.com/FerretDB/FerretDB/blob/9c440cc2d533508e619528cd1d605d5079daad7e/internal/handlers/commonparams/params.go#L84-L87

File: internal/handlers/common/params.go (hunk 3 above)
Comment (w84thesun) [reply internal/handlers/common/params.go:35]: It seems that I have to specify that it's not "errors" but error types. We can check that those errors have `CommandError` type but there is no way to check the error code now. We could possibly extend the `contract` package with `EnsureCommandError` and `EnsureWriteError`. What do you think?

Comment (AlekSi) [reply internal/handlers/common/params.go:35]: That's what I tried to do – and that's surprisingly not easy.

Let's only update documentation for now. `ErrBadValue` is not an error and also not a error type

**Review by rumyantseva — APPROVED** [review]

File: internal/handlers/common/params.go (hunk 3 above)
Comment (chilagrow) [reply internal/handlers/common/params.go:35]: I like this. I think this is the practice we can expand for new code bases :hugs:

**Review by chilagrow — APPROVED** [review]

**Comment by AlekSi** [comment]
On hold until after 1.2.0 release

**Review by rumyantseva — APPROVED** [review]
LGTM!

**Pull request merged**